File: Program's_Contributed_By_Contributors/C++_Programs/Sorting/SelectionSort.cpp
// Introduction to selection sort
// The algorithm divides input array into two parts 
// 1) Sorted Part which is built up from left to right
// 2) Unsorted part. Elements remaining to be sorted 
// Initially, sorted part is empty and unsorted part is the entire input array 

// Algorithm finds the smallest element and places in the sorted array at the end(right) in each iteraion
// For the first iteration it will place at 0th index  
// We can say that the algorithm places smallest element at ith position, where i = iteration count 

// Time Complexity of Selection Sort
// Time compelxity - O(n^2) in all cases which makes it ineffecient on large arrays 

// Space Complexity - O(1)


// C++ code for Selection Sort 
#include<iostream>
#include <bits/stdc++.h>
using namespace std;

// Function which will sort the array using Selection Sort algorithm
void selectionSort(int arr[],int n){
    // outer for loop
    // minimum element will be placed at ith position
    for(int i=0;i<n-1;i++){
        int min = i;
        for(int j=i+1;j<n;j++){
            // finding minimum element and storing its index in min
            if(arr[j]<arr[min]){
                min=j;
            }
        }
        // if minimum element is found it will be swapped with the ith positon
        if(min!=i){
            int temp = arr[min];
            arr[min]=arr[i];
            arr[i]= temp;
        }
    }
}

// main() function 
int main(){
    // creating array 
    int arr[] ={23,0,44,1,3,90,31,26,100};

    // getting length of array 
    int n =  sizeof(arr)/sizeof(arr[0]);

    // calling selectionSort() function
    // It takes an array and length of array as arguments 
    selectionSort(arr,n);

    //printing sorted array
    cout<<"\nSorted Array : ";
    for (int i = 0; i < n; i++)
    {
        cout<<arr[i]<<" ";
    }
    // Expected output : Sorted Array : 0 1 3 23 26 31 44 90 100 
    return 0;
}

File: Sorting/Selection_Sort.java
//Selection Sort repeatedly finds the minimum element and moves it to the beginning of the list.
//Time Complexity: O(n²)

// Selection Sort algorithm:

// Divide the list into two parts:

// The sorted part (initially empty).
// The unsorted part (the entire array at first).
// Find the minimum element:

// From the unsorted part, identify the smallest element.
// Swap:

// Swap the smallest element with the first element of the unsorted part.
// Move boundary:

// The first element of the unsorted part becomes part of the sorted portion.
// Repeat:

// Repeat steps 2-4 for the remaining unsorted part until the entire array is sorted.
// The process continues until all elements are sorted in ascending order.

//Code
public class Selection_Sort {
    static void selectionSort(int[] arr) {
        int n = arr.length;
        for (int i = 0; i < n - 1; i++) {
            int minIdx = i;
            for (int j = i + 1; j < n; j++) {
                if (arr[j] < arr[minIdx]) {
                    minIdx = j;
                }
            }
            // Swap the found minimum element with the first element
            int temp = arr[minIdx];
            arr[minIdx] = arr[i];
            arr[i] = temp;
        }
    }

    public static void main(String[] args) {
        int[] arr = {64, 25, 12, 22, 11};
        selectionSort(arr);
        System.out.println("Sorted array: ");
        for (int i : arr) {
            System.out.print(i + " ");
        }
    }
}
